llama_index/query_engine/retry_query_engine.py

```diff
@@ -0,0 +1,292 @@
+            query_bundle=query_bundle,
+            nodes=nodes,
+            additional_source_nodes=additional_source_nodes,
+        )
+
+    def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Answer a query."""
+        query_id = self.callback_manager.on_event_start(CBEventType.QUERY)
+
+        retrieve_id = self.callback_manager.on_event_start(CBEventType.RETRIEVE)
+        nodes = self._retriever.retrieve(query_bundle)
+        self.callback_manager.on_event_end(
+            CBEventType.RETRIEVE, payload={"nodes": nodes}, event_id=retrieve_id
+        )
+
+        synth_id = self.callback_manager.on_event_start(CBEventType.SYNTHESIZE)
+        response = self._response_synthesizer.synthesize(
+            query_bundle=query_bundle,
+            nodes=nodes,
+        )
+        self.callback_manager.on_event_end(
+            CBEventType.SYNTHESIZE, payload={"response": response}, event_id=synth_id
+        )
+
+        self.callback_manager.on_event_end(CBEventType.QUERY, event_id=query_id)
+
+        # Insert retry here
+        if self.max_retries > 0:
+            evaluator = QueryResponseEvaluator(service_context=self.service_context)
+
+            if type(response) is StreamingResponse:
+                typed_response = response.get_response()
+            elif type(response) is Response:
+                typed_response = response
+            else:
+                raise ValueError(f"Unexpected response type {type(response)}")
+
+            response_eval = evaluator.evaluate(query_bundle.query_str, typed_response)
+            if response_eval == "YES":
+                return response
+            else:
+                logger.debug(
+                    f">Binary evaluator returned NO, response: \
+                    {typed_response.response}"
+                )
+
+            old_source_nodes = evaluator.get_context(typed_response)
+            new_index = GPTListIndex.from_documents(old_source_nodes)
+            if self.check_source:
+                source_node_evals = evaluator.evaluate_source_nodes(
+                    query_bundle.query_str, typed_response
+                )
+                assert len(old_source_nodes) == len(source_node_evals)
+                new_source_docs: List[Document] = []
+                for i in range(len(source_node_evals)):
+                    if eval == "YES":
+                        new_source_docs.append(old_source_nodes[i])
+                    else:
+                        logger.debug(
+                            f">Source evaluator return NO, node: {old_source_nodes[i]}"
+                        )
+                if len(new_source_docs) == 0:
+                    # TODO: Add context-less query.
+                    logger.warn("No source nodes are relevant")
+                if new_source_docs:
+                    new_index = GPTListIndex.from_documents(new_source_docs)
+
+            new_retriever = new_index.as_retriever(
```

Comment:
maybe this is because i don't fully understand your design: i don't really understand why we're constructing a new index just over the source nodes. maybe because i don't fully get the assumptions being made in the retry mechanism. what types of errors is this supposed to help correct?

Reply:
I'm starting to understand your philosophy after digging through this more. But instead of constructing a list index, it feels like a cleaner abstraction to just construct a RetrieverQueryEngine with the new prompts and "re-query" everything. Yes that would also require "re-retrieving stuff"